### PR TITLE
ADR 1

### PR DESCRIPTION
**Pull request by deepansh96**
ETL Flow for homework type quizzes

### CODE REVIEW

File: docs/adrs/ADR-1.md [hunk 1]
```diff
@@ -1,5 +1,4 @@
 # ADR - 1
-# THIS IS TEMPORARY -- WIll be making a valid ADR soon
 
 ## Title
```
Comment (pritamps) [inline docs/adrs/ADR-1.md:3]: Maybe add a last-updated date here, so that people can know if things get very out of date. For example, if we start using a different Google Sheet, or if we switch to some UI -- we have to update those values in the ADRs. As the number of ADRs grow, I suspect this will get a bit hard.

File: docs/adrs/ADR-1.md [hunk 2]
```diff
@@ -13,29 +12,97 @@ Deepansh
...
+
 ## Context & Problem Description
 > Go into detail about the subject in question. Why is this decision important? What problems are there that this decision needs to be made? The urgency of the decision? Datapoints and related background information. Vocabulary and key terms.
 
+Till now we have only been syncing assessment sessions but not homework sessions. For homework sessions, the extracted data will come from two places: **Firestore Sessions DB** and **Quiz Mongo DB**.
+The data being extracted from the **Firestore Sessions DB** is the data pertaining to the sessions. Things like `program`, `grade`, `course`, `stream` and many other, basically the column values stored in the **Quiz Session Creator** google sheet. The data being extracted from the **Quiz Mongo DB** will be the responses of students who have attempted that quiz and the related metadata. Decisions regarding some questions mentioned below need to be taken:
+1. Where should the homework sessions data be stored on BQ? Should it be stored in the `test_responses` table that exists or some new table?
```
Comment (pritamps) [inline docs/adrs/ADR-1.md:27]: It depends on the use case completely. Not on tech simplicity or anything else. If we predict that analysis for tests and homeworks will be mostly done together, then they should be in the same table. If not, then separate.

Here my intuition is that they should be separate (even if columns are the same). Analysis will be *mostly* for tests separately and homeworks separately (I say *mostly*, because for things like student profile, we'll get both), so for the analysis this separation will be done with filters. Instead of this, it seems better to me to just keep the separation here itself and do the concatenation for the few use cases where it's needed

File: docs/adrs/ADR-1.md [hunk 3]
```diff
@@ -13,29 +12,97 @@ Deepansh
...
+2. If storing in a different table, how should we name the new table? The old one is `test_responses`. It doesn't have the keyword `assessments` in the name.
```
Comment (pritamps) [inline docs/adrs/ADR-1.md:28]: Isn't it `assessments/test_responses` on BigQuery? This can be `assessments/homework_responses`

File: docs/adrs/ADR-1.md [hunk 4]
```diff
@@ -13,29 +12,97 @@ Deepansh
...
+3. Is the question level analysis also needed for homework tests? If yes, then should it be stored in a separate table or in the table for `assessments`?
```
Comment (pritamps) [inline docs/adrs/ADR-1.md:29]: Yes, needed. Question level can be same table, because here the use case will be twofold:

1. Getting question level stuff in test/homework reports. In this case, we're querying by session ID, so it doesn't matter that they are in the same table
2. Evaluating a question: Here we will most likely analyse all historical responses to a question (both homework and test, and maybe even Plio) -- so makes sense for it to be in the same table

File: docs/adrs/ADR-1.md [hunk 5]
```diff
@@ -13,29 +12,97 @@ Deepansh
...
+4. This data also needs to go into the Dynamo DB (for reporting engine). Should it go to a different table or the same existing one?
```
Comment (pritamps) [inline docs/adrs/ADR-1.md:30]: This is a good question... Hmm... I think same table is fine. Since it's called `student_quiz_reports`, and `quiz` here indicates a session on our quiz engine, whether homework or test

File: docs/adrs/ADR-1.md [hunk 6]
```diff
@@ -13,29 +12,97 @@ Deepansh
...
+5. What all fields are necessary or good to have for the homework level sessions? The fields listed below are the ones present in the assessments table. We can filter out or add more fields for the homework quizzes. Decision needs to be taken on entries with a :question: emoji:
+
+    | Field | Comments |
+    | ------ | ----------- |
+    | user_id | will be added |
+    | num_skipped | :question: Right now one can't skip in homework mode. But it's possible in the future, so for now, should we keep it and the value will be 0? |
```
Comment (pritamps) [inline docs/adrs/ADR-1.md:36]: Yes yes, keep

File: docs/adrs/ADR-1.md [hunk 7]
```diff
@@ -13,29 +12,97 @@ Deepansh
...
+    | num_correct | :question: Number of correctly answered graded questions. This will not include ungraded questions right? |
```
Comment (pritamps) [inline docs/adrs/ADR-1.md:37]: Hmm... good question. You should talk to test people and see what they need. If a question has a correct/wrong AND it is ungraded, to me it makes sense to include it.

File: docs/adrs/ADR-1.md [hunk 8]
```diff
@@ -13,29 +12,97 @@ Deepansh
...
+    | num_wrong | :question: Number of wrongly answered graded questions. This will not include ungraded questions right? |
```
Comment (pritamps) [inline docs/adrs/ADR-1.md:38]: same as above

File: docs/adrs/ADR-1.md [hunk 9]
```diff
@@ -13,29 +12,97 @@ Deepansh
...
+    | marks_scored | :question: Marks scored by the user on graded questions wrt marking scheme. What is the default marking scheme if nothing is present? |
```
Comment (pritamps) [inline docs/adrs/ADR-1.md:39]: Is it possible for marking scheme to not be present in the current quiz creation API? If so, we should change that, even if it means putting a default in the *quiz engine*. The ETL flow should not assume anything here

File: docs/adrs/ADR-1.md [hunk 10]
```diff
@@ -13,29 +12,97 @@ Deepansh
...
+    | percentage | Marks scored by user / total marks possible |
+    | section | :question: Will these questions also have subject wise sections like P/C/M/B? |
```
Comment (pritamps) [inline docs/adrs/ADR-1.md:41]: Probably not, but maybe yes in the future.

File: docs/adrs/ADR-1.md [hunk 11]
```diff
@@ -13,29 +12,97 @@ Deepansh
...
+    | highest_test_score | highest of scores of users in a particular session |
+    | rank | rank of student among other students in a particular session |
+    | percentile | percentile of student among other students in a particular session |
+    | num_test_takers | number of unique users who attempted the quiz |
+    | session_id | unique session id |
+    | start_date | start date of session |
+    | start_time | start time of session |
+    | end_date | end date of session |
+    | end_time | end time of session |
+    | infinite_session | :question: Should we add this? if yes, then this needs to go into firestore sessions DB as well |
```
Comment (pritamps) [inline docs/adrs/ADR-1.md:54]: Yes, why not. Unless we're 100% sure infinite sessions will not exist for homework

File: docs/adrs/ADR-1.md [hunk 12]
```diff
@@ -13,29 +12,97 @@ Deepansh
...
+
+  6. A table like the above one needs to be created if we decide on saving question level analysis as well for homework quizzes.
+  7. Should the existing ETL Flow code for assessments be reused or a new code file be created for this?
+
 ## Solution(s)
 > If a solution has been discussed then that should be discussed here. If multiple solutions exist, and their discussion is not feasible/practical on discord, then discuss these potential alternatives and their impact.
 > What alternatives are being considered, their benefits, their costs (team resources, money, time frames), mitigations for any drawbacks. The team will collectively take a decision in the Pull Request itself. The agreed upon decision then should be documented in the PR.
 
+Proposed solutions for the questions above:
+  1. After writing this, I think this should be stored in the same `test_responses` table. As the data will be almost similar to the assessments data. The only difference being the `test_type` being `homework`.
```
Comment (pritamps) [inline docs/adrs/ADR-1.md:75]: Disagree. The decision of same table or not should be primarily based on the final use case (like I mentioned above). It can be exactly the same flow (since they are so similar) with the table name as just a parameter based on the session type

File: docs/adrs/ADR-1.md [hunk 13]
```diff
@@ -13,29 +12,97 @@ Deepansh
...
+  2. If we store in the same table, this question is redundant.
+  3. TBD
+  4. According to me, same table. Will keep things simpler to manage and the data also makes sense staying in the existing table.
```
Comment (pritamps) [inline docs/adrs/ADR-1.md:78]: Here agreed

File: docs/adrs/ADR-1.md [hunk 14]
```diff
@@ -13,29 +12,97 @@ Deepansh
...
+  5. TBD
+  6. TBD
+  7. Same flow. The current code skips the quizzes which are of `homework` type. The new change to the code will not skip `homework` type quizzes, will take them into consideration and will transform the data accordingly. And because of this, the name of the flow will have to be changed from `ETL Flow Assessments` to something like `ETL Flow Quizzes`.
```
Comment (pritamps) [inline docs/adrs/ADR-1.md:81]: Agreed

File: docs/adrs/ADR-1.md [hunk 15]
```diff
@@ -13,29 +12,97 @@ Deepansh
...
 > A decision may be "proposed" if the project stakeholders haven't agreed with it yet, or "accepted" once it is agreed. If a later ADR changes or reverses a decision, it may be marked as "deprecated" or "superseded" with a reference to its replacement. You can use any one of these given badges
 
 <Badge text="PROPOSED"/>
-<Badge text="ACCEPTED"/>
-<Badge text="DEPRECATED" type="error"/>
-<Badge text="SUPERSEDED"type="warning"/>
 
 ## Consequences
 > This section describes the resulting context, after applying the decision. All consequences should be listed here, not just the "positive" ones. A particular decision may have positive, negative, and neutral consequences, but all of them affect the team and project in the future.
+
```
Comment (pritamps) [inline docs/adrs/ADR-1.md:103]: Maybe add an extra paragraph here about how long ETL flow takes now, and how long you expect it to take when the Homework sessions are also added, since (I think?) there are more homeworks than tests?

**Review by pritamps — CHANGES_REQUESTED** [review]

File: docs/adrs/ADR-1.md (hunk 1 above)
Comment (deepansh96) [reply docs/adrs/ADR-1.md:3]: Makes sense! I was also apprehensive of adding the google sheet link as that might get outdated. I'll remove the link then.
If we add a UI, maybe that can come out as a separate small ADR

File: docs/adrs/ADR-1.md (hunk 3 above)
Comment (deepansh96) [reply docs/adrs/ADR-1.md:28]: There's some conflict about the terminology here I guess. The `assessments` is a quiz type, and `homework` is also a quiz type. The folder name is `assessments` currently and it has two tables - one for general overall stats and one for question level stats. If we stay consistent with the structure, the new table should come under a folder `homework` and the names will again be the same `homework/test_responses`.

So in a nutshell, this is how it might look like
```
assessments/
- test_responses
- test_responses_question_level

homework/
- test_responses
- test_responses_question_level
```

File: docs/adrs/ADR-1.md (hunk 4 above)
Comment (deepansh96) [reply docs/adrs/ADR-1.md:29]: Agreed

File: docs/adrs/ADR-1.md (hunk 7 above)
Comment (deepansh96) [reply docs/adrs/ADR-1.md:37]: Okay will gather views of Karn and Heena on this

File: docs/adrs/ADR-1.md (hunk 9 above)
Comment (deepansh96) [reply docs/adrs/ADR-1.md:39]: yes, marking_scheme has a value of `None` by default. So not required for the API. 
Makes sense, the default value should not be `None` but a default decided marking scheme

File: docs/adrs/ADR-1.md (hunk 10 above)
Comment (deepansh96) [reply docs/adrs/ADR-1.md:41]: should the code be written keeping that in mind? like for now, we can only have the `overall` section. and we can add more if needed in the future?

File: docs/adrs/ADR-1.md (hunk 11 above)
Comment (deepansh96) [reply docs/adrs/ADR-1.md:54]: Will talk to Karn and Heena

File: docs/adrs/ADR-1.md (hunk 15 above)
Comment (deepansh96) [reply docs/adrs/ADR-1.md:103]: How long as in? The execution time?

File: docs/adrs/ADR-1.md (hunk 3 above)
Comment (pritamps) [reply docs/adrs/ADR-1.md:28]: Ah I thin khaving it like

```
assessments/
- test_respones
- homework_responses
- all_responses_question_level 
```
makes more sense...

File: docs/adrs/ADR-1.md (hunk 10 above)
Comment (pritamps) [reply docs/adrs/ADR-1.md:41]: If the code is same as that for tests, all this will be automatic, no?

File: docs/adrs/ADR-1.md (hunk 15 above)
Comment (pritamps) [reply docs/adrs/ADR-1.md:103]: Haan execution time. Any other effects. Effect on lambda cost... whatever you can think of :D
